src/entities/interfaces/subscription-capable.ts

```diff
@@ -0,0 +1,111 @@
+import * as Subscription from '../../core/types/api/subscription';
+import { EventEmitCapable } from './event-emit-capable';
+
+/**
+ * SubscriptionCapable entity type.
+ *
+ * @internal
+ */
+export enum SubscriptionType {
+  /**
+   * Channel identifier, which is part of the URI path.
+   */
+  Channel,
+
+  /**
+   * Channel group identifiers, which is part of the query parameters.
+   */
+  ChannelGroup,
+}
+
+/**
+ * PubNub entity subscription configuration options.
+ */
+export type SubscriptionOptions = {
+  /**
+   * Whether presence events for an entity should be received or not.
+   */
+  receivePresenceEvents?: boolean;
+
+  /**
+   * Real-time event filtering function.
+   *
+   * Function can be used to filter out events which shouldn't be populated to the registered event listeners.
+   *
+   * **Note:** This function is called for each received event.
+   *
+   * @param event - Pre-processed event object from real-time subscription stream.
+   *
+   * @returns `true` if event should be populated to the event listeners.
+   */
+  filter?: (event: Subscription.SubscriptionResponse['messages'][0]) => boolean;
```

Comment:
Does this scale for all messages, just wondering how it filter through all messages?

Reply:
It is called on per-message basis. We don't have general names for those which is un `d` field, so sometime they are message and some time they are events. `event` here is a union of `*Event` types which user can review and return `true` if event should be handled.